### ISSUE TITLE
MTurk Single Agent Role Play

### ISSUE THREAD

**Issue by shinoys222** [facebookresearch/ParlAI#506]
The examples for mturk tasks always have more than 1 agent. Is there a functionality in parlai for the worker to switch sides in conversation from left and right chat box locations to provide both bot and user messages.?

**Comment by JackUrb** [comment]
Hi @shinoys222,

The qa_data_collection task to some degree only has one agent, but it does have scripted responses being created by the world itself which is occupying the other side of the chat. We don't support a feature for the mturk agent to have both sides of the conversation natively, but there certainly is a way to implement it in your task. We allow overriding our existing javascript to replace it with any desired functionality, so in this case you would only need to override the part that controls adding entered text to the display window.

It is important not to modify the agent_ids that are being sent back to your server, as those are used by the ParlAI-MTurk manager to verify that the messages have come from the mturk agent, but you should be able to implement exactly what you want by overriding the `add_message_to_conversation` function: 
https://github.com/facebookresearch/ParlAI/blob/5c890cdafc63267defda08934657b5109bb90dfb/parlai/mturk/core/server/html/core.html#L409

To do this you will need to add a `html` folder in your task folder, and a file inside called `<agent_id>_index.html`, where `<agent_id>` is the id that you give to an agent as in: https://github.com/facebookresearch/ParlAI/blob/5c890cdafc63267defda08934657b5109bb90dfb/parlai/mturk/tasks/qa_data_collection/run.py#L35

The following is a rough implementation of what that file should contain to make it work:
```
{% extends "core.html" %}
{% block additional_scripts %}
<script type="text/javascript">
var use_right_side_next_time = true;

function add_message_to_conversation(agent_id, message_text, is_self) {
  var float_loc = 'left';
  var alert_class = 'alert-warning';
  if (use_right_side_next_time) {
    float_loc = 'right';
    alert_class = 'alert-info';
    agent_id = 'CONVERSATION_ID_YOU_WANT_TO_DISPLAY_ON_THE_RIGHT';
    use_right_side_next_time = false;
  } else {
    agent_id = 'CONVERSATION_ID_YOU_WANT_TO_DISPLAY_ON_THE_LEFT';
    use_right_side_next_time = true;
  }
  $('div#message_thread').append(
    '<div class="row" style="margin-left: 0; margin-right: 0">' +
      '<div class="alert '+alert_class+'" role="alert" style="float: '+float_loc+'; display:table">' +
        '<span style="font-size: 16px"><b>'+agent_id+'</b>: '+message_text+'</span>' +
      '</div>'+
    '</div>'
  );
  $("div#message_thread").css("display", "");
  scroll_conversation_to_bottom();
}
</script>
{% endblock %}
```
This overrides just the `add_message_to_conversation` function, using a toggle variable (`use_right_side_next_time`) rather than the given inputs to determine what side of the conversation the text will appear on as well as the display name that will appear in the chat bubble. 

With this implementation, every time your ParlAI world calls `mturk_agent.act()` it will let the worker fill out one entry in the conversation. Eventually though the world must consider the conversation done and mark itself done as in: https://github.com/facebookresearch/ParlAI/blob/5c890cdafc63267defda08934657b5109bb90dfb/parlai/mturk/tasks/qa_data_collection/worlds.py#L71

Let me know if there's anything else I can do to help!

**Comment by JackUrb** [comment]
@shinoys222 any luck with the above implementation?

**Comment by shinoys222** [comment]
Yes @JackUrb thank you very much . Sorry, forgot to mention it before.

Also , please mention this in your documentation that, to check for disconnects  in parley function using `self.mturk_agent.disconnected`.Otherwise the parley function is being called forever during disconnects.
